linux_os/guide/system/network/network-firewalld/ruleset_modifications/ensure_firewall_rules_for_open_ports/oval/shared.xml

```diff
@@ -1,76 +1,300 @@
-      <linux:local_port datatype="int" operation="equals"  var_ref="var_firewalled_ports" var_check="at least one"/>
-    </linux:inetlisteningservers_state>
+  <linux:inetlisteningservers_test id="test_listening_fw_inet_direct_ports_tcp_test" version="1"
+                                   check="at least one"
+                                   comment="Check all listening ports defined by service are firewalled">
+    <linux:object object_ref="obj_listening_inet_tcp_ports" />
+    <linux:state state_ref="state_firewalled_listening_inet_tcp_ports" />
+  </linux:inetlisteningservers_test>
+
+  <ind:variable_test id="test_var_firewalled_direct_tcp_ports_exists" check="all"
+                     check_existence="at_least_one_exists" version="1"
+                     comment="Check the existence of direct tcp port definitions.">
+    <ind:object object_ref="object_var_firewalled_direct_tcp_ports"/>
+  </ind:variable_test>
+
+  <ind:variable_test id="test_var_firewalled_tcp_ports_exists" check="all"
+                     check_existence="at_least_one_exists" version="1"
+                     comment="Check the existence of tcp port defined through services.">
+    <ind:object object_ref="object_var_firewalled_tcp_ports"/>
+  </ind:variable_test>
 
-  <linux:inetlisteningservers_object id="obj_listening_inet_ports" version="2" comment="This object represents a listening services on the system.">
-      <linux:protocol operation="pattern match">^.*$</linux:protocol>
+  <linux:inetlisteningservers_state id="state_firewalled_listening_inet_tcp_ports" version="1"
+                                    comment="Checks listen ports has rule in firewalld">
+    <linux:local_port datatype="int" operation="equals"  var_ref="var_firewalled_tcp_ports"
+                      var_check="at least one"/>
+  </linux:inetlisteningservers_state>
+
+  <linux:inetlisteningservers_state id="state_firewalled_listening_inet_direct_tcp_ports" version="1"
+                                    comment="Checks listen ports has direct port rule in firewalld">
+    <linux:local_port datatype="int" operation="equals"  var_ref="var_firewalled_direct_tcp_ports"
+                      var_check="at least one"/>
+  </linux:inetlisteningservers_state>
+
+  <linux:inetlisteningservers_test id="test_listening_fw_inet_ports_tcp_exist" version="1" check="all"
+                                   check_existence="none_exist"
+                                   comment="Check if any service is listening on tcp ports">
+    <linux:object object_ref="obj_listening_inet_tcp_ports" />
+  </linux:inetlisteningservers_test>
+
+  <linux:inetlisteningservers_object id="obj_listening_inet_tcp_ports" version="2"
+                                     comment="Represents a listening services on the system.">
+      <linux:protocol operation="equals">tcp</linux:protocol>
       <linux:local_address operation="not equal">^.*$</linux:local_address>
```

Comment:
Is `not equal` correct here? It seems it would exclude everything (`^.*$`).

Reply:
Thanks for the feedback on this definitely looks messy. Not sure what was the idea there, maybe matching only those bound to ANY IP, which is obviously wrong. 
Anyways having in mind the filters I later included there the local_address comparison seems totally irrelevant there. 
As far as I can see from the filters and doing some manual tests with or without it, rule  behaves same, so dropping it out from the definition.